File: scripts/train_words.py
"""Latih model kata WL-BISINDO di TensorFlow Python (cepat), lalu ekspor
bobot ke JSON untuk disuntikkan ke arsitektur TFJS oleh inject_weights.mjs.

Pipeline identik dengan scripts/train_words.mjs dan notebook Kaggle:
fitur 48x36, split signer-independent, augmentasi mirror + jitter temporal.
"""

import json
import os

import numpy as np
import tensorflow as tf
from tensorflow.keras import layers

WINDOW = 48
FEAT = 36
EPOCHS = int(os.environ.get("EPOCHS", "120"))
ROOT = os.path.join(os.path.dirname(__file__), "..")

with open(os.path.join(ROOT, "data", "word_seqs.json")) as f:
    seqs = json.load(f)
with open(os.path.join(ROOT, "data", "word_labels.json")) as f:
    WORD_LABELS = json.load(f)

labels = sorted({s["label"] for s in seqs}, key=int)
label_idx = {l: i for i, l in enumerate(labels)}
words = [WORD_LABELS.get(l, f"kata-{l}") for l in labels]
print(f"{len(seqs)} sekuens, {len(labels)} kelas")


def resample(seq, n=WINDOW):
    seq = np.asarray(seq, dtype=np.float32)
    idx = np.linspace(0, len(seq) - 1, n)
    lo = np.floor(idx).astype(int)
    hi = np.minimum(lo + 1, len(seq) - 1)
    a = (idx - lo)[:, None]
    return seq[lo] * (1 - a) + seq[hi] * a


def mirror(frames):
    g = frames.copy()
    g[:, 0:15], g[:, 15:30] = frames[:, 15:30].copy(), frames[:, 0:15].copy()
    g[:, 30], g[:, 31] = frames[:, 31].copy(), frames[:, 30].copy()
    g[:, 32] = -frames[:, 34]
    g[:, 33] = frames[:, 35]
    g[:, 34] = -frames[:, 32]
    g[:, 35] = frames[:, 33]
    return g


def jitter(seq, rng):
    n = len(seq)
    cs = int(n * rng.random() * 0.12)
    ce = int(n * rng.random() * 0.12)
    sliced = seq[cs : n - ce]
    return resample(sliced) if len(sliced) >= 8 else resample(seq)


rng = np.random.default_rng(42)
signers = sorted({s["signer"] for s in seqs})
val_signer = signers[-1]
print(f"Signer: {signers}, validasi: {val_signer} (signer-independent)")

x_train, y_train, x_val, y_val = [], [], [], []
skipped = 0
for s in seqs:
    arr = np.asarray(s["seq"], dtype=np.float32)
    if len(arr) < 8 or not ((arr[:, 30] > 0) | (arr[:, 31] > 0)).any():
        skipped += 1
        continue
    base = resample(arr)
    yi = label_idx[s["label"]]
    if s["signer"] == val_signer:
        x_val.append(base)
        y_val.append(yi)
    else:
        x_train += [base, mirror(base), jitter(arr, rng), jitter(arr, rng)]
        y_train += [yi] * 4

x_train = np.stack(x_train)
y_train = np.array(y_train)
x_val = np.stack(x_val)
y_val = np.array(y_val)
print(f"Train {x_train.shape}, Val {x_val.shape}, dilewati {skipped}")

model = tf.keras.Sequential(
    [
        layers.Input(shape=(WINDOW, FEAT)),
        layers.Conv1D(64, 5, strides=2, activation="relu"),
        layers.Dropout(0.3),
        layers.Conv1D(128, 3, strides=2, activation="relu"),
        layers.Dropout(0.3),
        layers.GlobalAveragePooling1D(),
        layers.Dense(64, activation="relu"),
        layers.Dropout(0.2),
        layers.Dense(len(labels), activation="softmax"),
    ]
)
model.compile(
    optimizer=tf.keras.optimizers.Adam(1e-3),
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"],
)

best_path = os.path.join(ROOT, "data", "best_words.keras")
callbacks = [
    tf.keras.callbacks.ModelCheckpoint(
        best_path, monitor="val_accuracy", save_best_only=True, verbose=0
    ),
    tf.keras.callbacks.ReduceLROnPlateau(
        monitor="val_accuracy", factor=0.5, patience=10, min_lr=1e-5
    ),
]
model.fit(
    x_train,
    y_train,
    validation_data=(x_val, y_val),
    epochs=EPOCHS,
    batch_size=32,
    shuffle=True,
    callbacks=callbacks,
    verbose=2,
)

model = tf.keras.models.load_model(best_path)
val_loss, val_acc = model.evaluate(x_val, y_val, verbose=0)
print(f"\nAkurasi validasi terbaik (signer-independent): {val_acc * 100:.2f}%")

preds = model.predict(x_val, verbose=0).argmax(-1)
print("Akurasi per kata:")
for i, w in enumerate(words):
    mask = y_val == i
    if mask.any():
        ok = int((preds[mask] == i).sum())
        print(f"  {w}: {ok / mask.sum() * 100:.0f}% ({ok}/{int(mask.sum())})")

# ekspor bobot untuk disuntikkan ke tfjs (urutan = model.get_weights())
export = {
    "words": words,
    "valAccuracy": float(val_acc),
    "valSigner": val_signer,
    "weights": [w.tolist() for w in model.get_weights()],
}
out = os.path.join(ROOT, "data", "words_weights.json")
with open(out, "w") as f:
    json.dump(export, f)
print(f"Bobot terekspor ke {out}")

File: scripts/train_words.mjs
/*
 * Latih model pengenalan kata WL-BISINDO dari sekuens fitur landmark.
 * Input [48 frame x 36 fitur], model Conv1D kecil, disimpan sebagai
 * TFJS Layers ke public/models/bisindo-words/.
 * Split signer-independent: satu signer disisihkan untuk validasi.
 */

import fs from "node:fs";
import path from "node:path";
import * as tf from "@tensorflow/tfjs";

const WINDOW = 48;
const FEAT = 36;
const dataDir = path.join(process.cwd(), "data");
const outDir = path.join(process.cwd(), "public", "models", "bisindo-words");

const seqs = JSON.parse(
  fs.readFileSync(path.join(dataDir, "word_seqs.json"), "utf8"),
);
const labelWords = JSON.parse(
  fs.readFileSync(path.join(dataDir, "word_labels.json"), "utf8"),
);

const labels = [...new Set(seqs.map((s) => s.label))].sort(
  (a, b) => Number(a) - Number(b),
);
const labelIdx = new Map(labels.map((l, i) => [l, i]));
const words = labels.map((l) => labelWords[l] ?? `kata-${l}`);
console.log(`${seqs.length} sekuens, ${labels.length} kelas`);

function resample(seq, n) {
  const out = [];
  for (let i = 0; i < n; i++) {
    const t = (i / (n - 1)) * (seq.length - 1);
    const lo = Math.floor(t);
    const hi = Math.min(seq.length - 1, lo + 1);
    const a = t - lo;
    const row = new Array(FEAT);
    for (let f = 0; f < FEAT; f++) {
      row[f] = seq[lo][f] * (1 - a) + seq[hi][f] * a;
    }
    out.push(row);
  }
  return out;
}

function mirror(frames) {
  // tukar tangan kiri-kanan dan balik sumbu x wrist
  return frames.map((f) => {
    const g = new Array(FEAT).fill(0);
    for (let i = 0; i < 15; i++) {
      g[i] = f[15 + i];
      g[15 + i] = f[i];
    }
    g[30] = f[31];
    g[31] = f[30];
    g[32] = -f[34];
    g[33] = f[35];
    g[34] = -f[32];
    g[35] = f[33];
    return g;
  });
}

function jitter(frames) {
  // pangkas acak 0-12% dari awal/akhir lalu resample lagi
  const cutStart = Math.floor(frames.length * Math.random() * 0.12);
  const cutEnd = Math.floor(frames.length * Math.random() * 0.12);
  const sliced = frames.slice(cutStart, frames.length - cutEnd);
  return sliced.length >= 8 ? resample(sliced, WINDOW) : frames;
}

const signers = [...new Set(seqs.map((s) => s.signer))].sort();
const valSigner = signers[signers.length - 1];
console.log(`Signer: ${signers.join(", ")}, validasi: ${valSigner}`);

const train = { x: [], y: [] };
const val = { x: [], y: [] };
let skipped = 0;
for (const s of seqs) {
  const hasHands = s.seq.some((f) => f[30] > 0 || f[31] > 0);
  if (!hasHands || s.seq.length < 8) {
    skipped++;
    continue;
  }
  const base = resample(s.seq, WINDOW);
  const yi = labelIdx.get(s.label);
  if (s.signer === valSigner) {
    val.x.push(base);
    val.y.push(yi);
  } else {
    train.x.push(base);
    train.y.push(yi);
    train.x.push(mirror(base));
    train.y.push(yi);
    train.x.push(jitter(base));
    train.y.push(yi);
  }
}
console.log(
  `Train ${train.x.length} (dengan augmentasi), Val ${val.x.length}, dilewati ${skipped}`,
);

const xTrain = tf.tensor3d(train.x);
const yTrain = tf.oneHot(tf.tensor1d(train.y, "int32"), labels.length);
const xVal = tf.tensor3d(val.x);
const yVal = tf.oneHot(tf.tensor1d(val.y, "int32"), labels.length);

/* Checkpoint berkala: simpan bobot tiap CKPT_EVERY epoch ke folder
 * checkpoint, dan lanjutkan dari sana bila run sebelumnya terputus. */
const CKPT_EVERY = 5;
const ckptDir = path.join(outDir, "checkpoint");

function saveHandlerTo(dir, extraMeta = {}) {
  return tf.io.withSaveHandler(async (artifacts) => {
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(
      path.join(dir, "weights.bin"),
      Buffer.from(artifacts.weightData),
    );
    fs.writeFileSync(
      path.join(dir, "model.json"),
      JSON.stringify({
        modelTopology: artifacts.modelTopology,
        format: "layers-model",
        generatedBy: "tfjs-node-train",
        convertedBy: null,
        weightsManifest: [
          { paths: ["./weights.bin"], weights: artifacts.weightSpecs },
        ],
      }),
    );
    if (Object.keys(extraMeta).length) {
      fs.writeFileSync(
        path.join(dir, "state.json"),
        JSON.stringify(extraMeta),
      );
    }
    return {
      modelArtifactsInfo: { dateSaved: new Date(), modelTopologyType: "JSON" },
    };
  });
}

async function loadCheckpoint() {
  const mj = path.join(ckptDir, "model.json");
  const st = path.join(ckptDir, "state.json");
  if (!fs.existsSync(mj) || !fs.existsSync(st)) return null;
  const state = JSON.parse(fs.readFileSync(st, "utf8"));
  const modelJson = JSON.parse(fs.readFileSync(mj, "utf8"));
  const weights = fs.readFileSync(path.join(ckptDir, "weights.bin"));
  const m = await tf.loadLayersModel(
    tf.io.fromMemory({
      modelTopology: modelJson.modelTopology,
      weightSpecs: modelJson.weightsManifest[0].weights,
      weightData: new Uint8Array(weights).buffer,
    }),
  );
  console.log(`Checkpoint ditemukan di epoch ${state.epoch}, dilanjutkan`);
  return { model: m, doneEpochs: state.epoch };
}

const resumed = await loadCheckpoint();

const model = resumed?.model ?? tf.sequential({
  layers: [
    tf.layers.conv1d({
      inputShape: [WINDOW, FEAT],
      filters: 64,
      kernelSize: 5,
      strides: 2,
      activation: "relu",
    }),
    tf.layers.dropout({ rate: 0.3 }),
    tf.layers.conv1d({ filters: 128, kernelSize: 3, strides: 2, activation: "relu" }),
    tf.layers.dropout({ rate: 0.3 }),
    tf.layers.globalAveragePooling1d({}),
    tf.layers.dense({ units: 64, activation: "relu" }),
    tf.layers.dropout({ rate: 0.2 }),
    tf.layers.dense({ units: labels.length, activation: "softmax" }),
  ],
});
model.compile({
  optimizer: tf.train.adam(1e-3),
  loss: "categoricalCrossentropy",
  metrics: ["accuracy"],
});

const TOTAL_EPOCHS = Number(process.env.EPOCHS ?? 60);
const startEpoch = resumed?.doneEpochs ?? 0;
await model.fit(xTrain, yTrain, {
  epochs: TOTAL_EPOCHS - startEpoch,
  batchSize: 32,
  shuffle: true,
  validationData: [xVal, yVal],
  callbacks: {
    onEpochEnd: async (e, logs) => {
      const epoch = startEpoch + e + 1;
      if (epoch % 1 === 0)
        console.log(
          `epoch ${epoch}: loss=${logs.loss.toFixed(3)} acc=${logs.acc.toFixed(3)} val_acc=${logs.val_acc.toFixed(3)}`,
        );
      if (epoch % CKPT_EVERY === 0 && epoch < TOTAL_EPOCHS) {
        await model.save(saveHandlerTo(ckptDir, { epoch, valAcc: logs.val_acc }));
      }
    },
  },
});

const evalRes = model.evaluate(xVal, yVal, { batchSize: 128 });
const valAcc = (await evalRes[1].data())[0];
console.log(`\nAkurasi validasi signer-independent: ${(valAcc * 100).toFixed(2)}%`);

// akurasi per kata
const preds = model.predict(xVal).argMax(-1);
const predArr = await preds.data();
const per = words.map(() => ({ ok: 0, total: 0 }));
val.y.forEach((yi, i) => {
  per[yi].total++;
  if (predArr[i] === yi) per[yi].ok++;
});
console.log("Akurasi per kata (validasi):");
per.forEach((s, i) => {
  if (s.total)
    console.log(`  ${words[i]}: ${((s.ok / s.total) * 100).toFixed(0)}% (${s.ok}/${s.total})`);
});

fs.mkdirSync(outDir, { recursive: true });
await model.save(saveHandlerTo(outDir));
// checkpoint tidak diperlukan lagi setelah model final tersimpan
fs.rmSync(ckptDir, { recursive: true, force: true });
fs.writeFileSync(
  path.join(outDir, "metadata.json"),
  JSON.stringify(
    {
      words,
      window: WINDOW,
      featureSize: FEAT,
      valAccuracy: valAcc,
      valSigner,
      dataset: "kaggle:glennleonali/wl-bisindo (CC BY-NC 4.0)",
      trainedAt: new Date().toISOString(),
    },
    null,
    2,
  ),
);
console.log(`Model tersimpan di ${outDir}`);
